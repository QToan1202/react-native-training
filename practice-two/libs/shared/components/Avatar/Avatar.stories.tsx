import { ComponentMeta, ComponentStory } from '@storybook/react-native'

import { Avatar } from '../index'

export default {
  title: 'components/Avatar',
  component: Avatar,
} as ComponentMeta<typeof Avatar>

const Template: ComponentStory<typeof Avatar> = (args) => <Avatar {...args} />

export const Default = Template.bind({})
Default.args = {
  source: require('@assets/avatar.png'),
  name: 'avatar',
}

export const AvatarWithDiffSize = Template.bind({})
AvatarWithDiffSize.args = {
  ...Default.args,
  size: 'md',
}

export const AvatarFullOtp = Template.bind({})
AvatarFullOtp.args = {
  ...Default.args,
  size: 'xl',
  inline: true,
}
