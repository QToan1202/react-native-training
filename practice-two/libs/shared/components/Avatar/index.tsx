import { memo, useMemo, useRef } from 'react'
import isEqual from 'react-fast-compare'
import { getTokens, Circle, SizeTokens } from 'tamagui'
import { ImageStyle, StyleProp } from 'react-native'

import Paragraph from '../Paragraph'

import StyledStack, { StyledImage, StyledImageProps, StyledStackProps } from './styles'

export type AvatarProps = StyledStackProps & {
  source: StyledImageProps['source']
  name: string
  size: SizeTokens
}

const Avatar = ({ source, name, size = 'sm', ...rest }: AvatarProps) => {
  const imageSize = useRef<number>(getTokens().avatar[size].val)
  const imageStyle = useMemo<StyleProp<ImageStyle>>(
    () => ({
      width: imageSize.current,
      height: imageSize.current,
      borderRadius: 50,
    }),
    []
  )

  return (
    <StyledStack space={10} {...rest}>
      <Circle>
        <StyledImage source={source} style={imageStyle} accessibilityLabel={`${name}-avatar`} />
      </Circle>
      <Paragraph
        numberOfLines={1}
        color="$color.gray_50"
        textTransform="capitalize"
        lineHeight="$3"
        fontWeight="$2"
        content={name}
      />
    </StyledStack>
  )
}

export default memo(Avatar, isEqual)
