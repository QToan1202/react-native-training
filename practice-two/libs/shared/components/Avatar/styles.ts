import { Image } from 'react-native'
import { GetProps, Stack, styled } from 'tamagui'

const StyledStack = styled(Stack, {
  space: 10,
  alignItems: 'center',
  columnGap: 4,

  variants: {
    inline: {
      true: {
        flexDirection: 'row',
      },
    },

    block: {
      true: {
        flexDirection: 'column',
      },
    },
  } as const,

  defaultVariants: {
    inline: true,
  },
})

export type StyledStackProps = GetProps<typeof StyledStack>

export const StyledImage = styled(Image, {
  name: 'Image',
  source: { uri: '' },
})

export type StyledImageProps = GetProps<typeof StyledImage>

export default StyledStack
