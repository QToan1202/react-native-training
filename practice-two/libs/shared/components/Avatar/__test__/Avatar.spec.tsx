import { render } from '@testing-library/react-native'

import Avatar from '../index'

describe('Testing avatar component', () => {
  it('Render avatar successfully', () => {
    const component = render(<Avatar source={require('@assets/avatar.png')} name="user" />).toJSON()

    expect(component).toMatchSnapshot()
  })

  it('Avatar must have accessibilityLabel that include name prop', () => {
    const avatarName: string = 'user'
    const { queryByLabelText } = render(
      <Avatar source={require('@assets/avatar.png')} name={avatarName} testID="avatar" />
    )

    expect(queryByLabelText(`${avatarName}-avatar`)).toBeTruthy()
    expect(queryByLabelText(`${avatarName}`)).toBeFalsy()
    expect(queryByLabelText('avatar')).toBeFalsy()
  })

  it('Avatar component have to contain Image and Paragraph component', () => {
    const { getByTestId, getByText, getByLabelText } = render(
      <Avatar source={require('@assets/avatar.png')} name="user" testID="avatar" />
    )
    const ImageComponent = getByLabelText('user-avatar')
    const ParagraphComponent = getByText('user')

    expect(getByTestId('avatar')).toContainElement(ImageComponent)
    expect(getByTestId('avatar')).toContainElement(ParagraphComponent)
  })
})
