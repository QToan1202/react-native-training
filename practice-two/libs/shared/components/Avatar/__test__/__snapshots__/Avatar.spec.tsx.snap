// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing avatar component Render avatar successfully 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "columnGap": 4,
        "flexDirection": "row",
      },
      {
        "columnGap": 10,
      },
      undefined,
    ]
  }
>
  <Image
    accessibilityLabel="user-avatar"
    source={
      {
        "testUri": "../../../assets/avatar.png",
      }
    }
    style={
      [
        {
          "borderRadius": 50,
        },
        {
          "height": 20,
          "width": 20,
        },
      ]
    }
  />
  <Text
    numberOfLines={1}
    style={
      [
        {
          "color": "#fff",
          "fontFamily": "Montserrat_400Regular",
        },
        {
          "fontSize": 14,
        },
        {
          "color": "#4f4f4f",
          "fontFamily": "Montserrat_500Medium",
          "textTransform": "capitalize",
        },
      ]
    }
  >
    user
  </Text>
</View>
`;
